src/fromager/commands/bootstrap.py

```diff
@@ -189,64 +197,83 @@ def write_constraints_file(
+            resolved[dep_name] = versions[0]
+            # Remove from unresolved dependencies list
+            unresolved_dependencies.remove((dep_name, nodes))
+
+    # Below this point we have built multiple versions of the same thing, so
+    # we need to try to determine if any one of those versions meets all of
+    # the requirements.
+
+    # Track which versions can be used by which parent requirement.
+    usable_versions: dict[str, list[str]] = {}
+    # Track how many total users of a requirement (by name) there are so we
+    # can tell later if any version can be used by all of them.
+    user_counter = 0
+
+    # Flag to see if something is resolved
+    resolved_something = True
+
+    # Outer while loop to resolve remaining dependencies with multiple versions
+    while unresolved_dependencies and resolved_something:
+        resolved_something = False
+        # Make copy of the original list
+        items_to_resolve = unresolved_dependencies[:]
+        for dep_name, nodes in items_to_resolve:
+            # Which parent requirements can use which versions of the dependency we
+            # are working on?
+            for node in nodes:
+                parent_edges = node.get_incoming_install_edges()
+                for parent_edge in parent_edges:
+                    parent_name = parent_edge.destination_node.canonicalized_name
+                    if (
+                        parent_name in resolved
+                        and resolved[parent_name]
+                        != parent_edge.destination_node.version
+                    ):
+                        continue
+                    for matching_version in parent_edge.req.specifier.filter(versions):
+                        usable_versions.setdefault(str(matching_version), []).append(
+                            str(parent_edge.destination_node.version)
+                        )
+                    user_counter += 1
+
+            # Look for one version that can be used by all the parent dependencies
+            # and output that if we find it. Otherwise, include a warning and report
+            # all versions so a human reading the file can make their own decision
+            # about how to resolve the conflict.
+            for v, users in usable_versions.items():
+                if len(users) == user_counter:
+                    version_strs = [str(v) for v in sorted(versions)]
+                    logger.debug(
+                        "%s: selecting %s from multiple candidates %s",
+                        dep_name,
+                        v,
+                        version_strs,
                     )
+                    # Set the flag to true since we were able to resolve something
+                    resolved[dep_name] = v
+            # resolved_something = True                  # This causes the infinite loop. Trying this inside if after line 255 also causes infinite loop
```

Comment:
This should definitely be indented inside the if like line 255. Maybe some of the other changes I mention above about initializing the variables will fix the loop?